app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java

```diff
@@ -136,125 +156,297 @@ public void setStudyIdIndex(DynamoIndexHelper studyIdIndex) {
-                            // No max length, or max length is longer than what we can fit in Synapse. This has to
-                            // be an attachment.
-                            return UploadFieldType.ATTACHMENT_BLOB;
-                        } else {
-                            // Short string. Can be a string.
-                            return UploadFieldType.STRING;
-                        }
-                    } else {
-                        // Constraints aren't StringConstraints, so we can't determine max length (if any). To be
-                        // safe, make this an attachment.
-                        return UploadFieldType.ATTACHMENT_BLOB;
-                    }
-                case INTEGER:
-                    return UploadFieldType.INT;
-                case DECIMAL:
-                    return UploadFieldType.FLOAT;
-                case BOOLEAN:
-                    return UploadFieldType.BOOLEAN;
-                case DATE:
-                    return UploadFieldType.CALENDAR_DATE;
-                case TIME:
-                    // Upload schema has no concept of local time. Accept a string.
-                    return UploadFieldType.STRING;
-                case DATETIME:
-                    return UploadFieldType.TIMESTAMP;
-                default:
-                    // Again, when in doubt, default to inline_json_blob.
-                    return UploadFieldType.INLINE_JSON_BLOB;
+            // Get upload field type from the map.
+            uploadFieldType = SURVEY_TO_SCHEMA_TYPE.get(surveyQuestionType);
+            if (uploadFieldType == null) {
+                throw new BridgeServiceException("Unexpected survey question type: " + surveyQuestionType);
+            }
+
+            // Type-specific parameters.
+            if (constraints instanceof StringConstraints) {
+                Integer maxLength = ((StringConstraints) constraints).getMaxLength();
+                if (maxLength != null) {
+                    fieldDefBuilder.withMaxLength(maxLength);
+                } else {
+                    // No max length specified. Assume this can be unbounded.
+                    fieldDefBuilder.withUnboundedText(true);
+                }
+            }
+        }
+
+        fieldDefBuilder.withType(uploadFieldType);
+        fieldDefList.add(fieldDefBuilder.build());
+
+        // NumericalConstraints (integer, decimal, duration) have units. We want to write the unit into Synapse in case
+        // (a) the survey question changes the units or (b) we add support for app-specified units.
+        if (constraints instanceof NumericalConstraints) {
+            UploadFieldDefinition unitFieldDef = new DynamoUploadFieldDefinition.Builder()
+                    .withName(fieldName + UploadUtil.UNIT_FIELD_SUFFIX).withType(UploadFieldType.STRING)
+                    .withRequired(false).withMaxLength(Unit.MAX_STRING_LENGTH).build();
+            fieldDefList.add(unitFieldDef);
```

Comment:
When you create an Int question "foo", we create schema fields "foo" and "foo_unit".